README.md
# Splunk HTTP Event Collector Stream for Bunyan

#### Version 0.11.0

This project provides a [Bunyan](https://www.npmjs.com/package/bunyan) stream for HTTP Event Collector in Splunk Enterprise and Splunk Cloud.

## Requirements

* Node.js v4 or later. Splunk HTTP Event Collector Stream for Bunyan is tested with Node.js v10.0 and v14.0.
* Splunk Enterprise 6.3.0 or later, or Splunk Cloud. Splunk HTTP Event Collector Stream for Bunyan is tested with Splunk Enterprise 8.0 and 8.2.0.
* An HTTP Event Collector token from your Splunk Enterprise or Splunk Cloud server.
* [Bunyan](https://www.npmjs.com/package/bunyan) (`npm install --save bunyan`).

## Installation

First, update npm to the latest version by running: `sudo npm install npm -g`.

Then run: `npm install --save splunk-bunyan-logger`.

## Usage

See the `examples` folder for usage examples:

* `all_batching.js`: Shows how to configure a Bunyan Stream with the 3 batching settings: `batchInterval`, `maxBatchCount`, & `maxBatchSize`.
* `basic.js`: Shows how to configure a Bunyan stream and send a log message to Splunk.
* `custom_format.js`: Shows how to configure a Bunyan Stream to log messages to Splunk using a custom format.
* `manual_batching.js`: Shows how to queue log messages, and send them in batches by manually calling `flush()`.
* `retry.js`: Shows how to configure retries on errors.

### SSL

Note: SSL certificate validation is diabled by default.
To enable it, set `logger.requestOptions.strictSSL = true` on your `SplunkStream` instance:

```javascript
var bunyan = require("bunyan");
var splunkBunyan = require("splunk-bunyan-logger");

var config = {
    token: "your-token-here",
    url: "https://splunk.local:8088"
};

var splunkStream = splunkBunyan.createStream(config);
// Enable SSL certificate validation
stream.logger.requestOptions.strictSSL = true;

// Note: splunkStream must be set to an element in the streams array
var Logger = bunyan.createLogger({
    name: "my logger",
    streams: [
        splunkStream
    ]
});
```

### Basic example

```javascript
var bunyan = require("bunyan");
var splunkBunyan = require("splunk-bunyan-logger");

var config = {
    token: "your-token-here",
    url: "https://splunk.local:8088"
};

var splunkStream = splunkBunyan.createStream(config);

// Note: splunkStream must be set to an element in the streams array
var Logger = bunyan.createLogger({
    name: "my logger",
    streams: [
        splunkStream
    ]
});

var payload = {
    // Message can be anything; doesn't have to be an object
    message: {
        temperature: "70F",
        chickenCount: 500
    }
};

console.log("Sending payload", payload);
Logger.info(payload, "Chicken coup looks stable.");
```

## Community

Stay connected with other developers building on Splunk software.

<table>

<tr>
<td><b>Email</b></td>
<td>devinfo@splunk.com</td>
</tr>

<tr>
<td><b>Issues</b>
<td><span>https://github.com/splunk/splunk-bunyan-logger/issues/</span></td>
</tr>

<tr>
<td><b>Answers</b>
<td><span>http://answers.splunk.com/</span></td>
</tr>

<tr>
<td><b>Blog</b>
<td><span>http://blogs.splunk.com/dev/</span></td>
</tr>

<tr>
<td><b>Twitter</b>
<td>@splunkdev</td>
</tr>

</table>

### Contact us

You can [contact support][contact] if you have Splunk related questions.

You can reach the developer platform team at _devinfo@splunk.com_.

## License

The Splunk HTTP Event Collector Stream for Bunyan is licensed under the Apache
License 2.0. Details can be found in the LICENSE file.

[contact]:                  https://www.splunk.com/en_us/support-and-services.html


CHANGELOG.md
# Splunk HTTP Event Collector Stream for Bunyan

## v0.11.0

### New features & APIs

* Replaced client implementation to use `needle` in lieu of deprecated `request`.
* Replaced client implementation to use `nyc` in lieu of deprecated `istanbul`.

### Minor changes

* Converted CI from Travis to Github Actions and added CD
* Upgrade version of bunyan to 1.8.15.
* Upgrade version of splunk-logging to 0.11.1.
* Upgrade version of jsdoc to 3.6.7.
* Upgrade version of jshint to 2.12.0.
* Upgrade version of mocha to 8.4.0.

## v0.10.1

### Breaking changes

* Dropped support for Node.js versions older than 4.0.0.

## v0.9.3

### Minor changes

* Update version of `splunk-logging` dependency, allowing new installs to receive security updates to the transitive `request` dependency.

## v0.9.2

### Bug Fixes

* Workaround a urlencoding bug in the request library (GitHub issue [#6](https://github.com/splunk/splunk-javascript-logging/issues/6) for the underlying `splunk-logging` library).
* Catch JSON parsing errors from the server without crashing (GitHub issue [#9](https://github.com/splunk/splunk-javascript-logging/issues/9) for the underlying `splunk-logging` library).


## v0.9.1

### Bug Fixes

* Relax port validation for ports < 1000.

## v0.9.0

### New Features & APIs

* Added the ability to configure automated batching with 3 settings: `batchInterval`, `maxBatchCount`, & `maxBatchSize`.
* Added the ability to retry sending to Splunk in the case of network errors with the `maxRetries` configuration setting.
* Added the ability to configure a custom Splunk event format by overriding `eventFormatter(message, severity)`.

### Breaking Changes

* Removed the `autoFlush` configuration setting. To achieve the same effect, set `config.maxBatchCount` to `0`.
* Removed support for middleware functions.
* The `context` object has been simplified, `config` and `requestOptions` can no longer be specified there - please use those settings directly on the logger.

### Examples

* Removed the `middleware.js` example.
* Renamed the `batching.js` example to `manual_batching`.
* Added the `all_batching.js`, `custom_format.js`, `retry.js` examples.

### Minor changes

* Significant refactor of internal functions.

## v0.8.0 - beta

* Beta release.

test/test_config.js
/*
 * Copyright 2015 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
var splunkBunyan = require("../index");
var assert = require("assert");

/** Unit tests **/

describe("createStream", function() {
    it("should error without config", function() {
        try {
            splunkBunyan.createStream();
            assert.ok(false, "Expected an error.");
        }
        catch(err) {
            assert.ok(err);
            assert.strictEqual(err.message, "Config is required.");
        }
    });
    it("should error without config", function() {
        var config = {};
        try {
            splunkBunyan.createStream(config);
            assert.ok(false, "Expected an error.");
        }
        catch(err) {
            assert.ok(err);
            assert.strictEqual(err.message, "Config object must have a token.");
        }
    });
    it("should create logger with minimal config", function() {
        var config = {
            token: "a-token-goes-here-usually"
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxBatchSize);
        assert.strictEqual(1, splunkBunyanStream.stream.config().maxBatchCount);
    });
    it("should create logger with manual batching enabled", function() {
        var config = {
            token: "a-token-goes-here-usually"
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxBatchSize);
        assert.strictEqual(1, splunkBunyanStream.stream.config().maxBatchCount);
    });
    it("should create logger with non-default name", function() {
        var config = {
            token: "a-token-goes-here-usually",
            name: "different name"
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual(config.name, splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
    });
    it("should create logger with non-default path", function() {
        var config = {
            token: "a-token-goes-here-usually",
            path: "/services/collector/different/1.0"
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual(config.path, splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
    });
    it("should create logger with maxRetries=5", function() {
        var config = {
            token: "a-token-goes-here-usually",
            maxRetries: 5
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(5, splunkBunyanStream.stream.config().maxRetries);
    });
    it("should create logger with maxBatchSize=100", function() {
        var config = {
            token: "a-token-goes-here-usually",
            maxBatchSize: 100
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
        assert.strictEqual(100, splunkBunyanStream.stream.config().maxBatchSize);
    });
    it("should create logger with maxBatchCount=10", function() {
        var config = {
            token: "a-token-goes-here-usually",
            maxBatchCount: 10
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.ok(splunkBunyanStream.hasOwnProperty("stream"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("logger"));
        assert.ok(splunkBunyanStream.stream.logger.hasOwnProperty("config"));
        assert.ok(splunkBunyanStream.stream.hasOwnProperty("send"));
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("/services/collector/event/1.0", splunkBunyanStream.stream.config().path);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxBatchSize);
        assert.strictEqual(10, splunkBunyanStream.stream.config().maxBatchCount);
    });
});

test/test_stream.js
/*
 * Copyright 2015 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

var splunkBunyan = require("../index");
var assert = require("assert");
var needle = require("needle");

/** Integration Tests **/

var TOKEN;

var successBody = {
    text: "Success",
    code: 0
};

var invalidTokenBody = {
    text: "Invalid token",
    code: 4
};

var ____consoleLog = console.log;
function mute() {
    console.log = function(){};
}
function unmute() {
    console.log = ____consoleLog;
}

function formatForBunyan(data) {
    var ret = {
        name: "a bunyan logger",
        hostname: "Shaqbook-15.local",
        pid: 37509,
        level: 30,
        msg: "",
        time: Date.now() / 1000, // The / 1000 part is for Splunk
        v: 0
    };

    if (typeof data === "string") {
        ret.msg = data;
    }
    else {
        for (var key in data) {
            ret[key] = data[key];
        }
    }

    return ret;
}

describe("Setup Splunk on localhost:8089 HEC", function() {
    it("should be enabled", function(done) {
        needle.post("https://admin:changed!@localhost:8089/servicesNS/admin/splunk_httpinput/data/inputs/http/http/enable?output_mode=json", null, {rejectUnauthorized: false}, function(err)     {
            assert.ok(!err);
            done();
        });
    });
    it("should create a token in test/config.json", function(done) {
        needle.post("https://admin:changed!@localhost:8089/servicesNS/admin/splunk_httpinput/data/inputs/http?output_mode=json", { name : "splunk_logging" + Date.now()}, {rejectUnauthorized: false}, function(err, resp, body) {
            assert.ok(!err);
            body = JSON.stringify(body);
            var tokenStart = body.indexOf("\"token\":\"");
            var tokenEnd = tokenStart + 36; // 36 = guid length
            var token = body.substring(tokenStart + 9, tokenEnd + 9); // 9 = prefix length of \"token\":\"
            assert.strictEqual(token.length, 36);
            TOKEN = token;
            done();
         });
    });
    it("should have the env variable set", function() {
        assert.ok(TOKEN);
        assert.strictEqual(TOKEN.length, 36);
    });
});

describe("SplunkStream", function() {

    it("should write a string", function(done) {
        var config = {
            token: TOKEN,        };

        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxRetries);
        assert.strictEqual(1, splunkBunyanStream.stream.config().maxBatchCount);
        assert.strictEqual(0, splunkBunyanStream.stream.config().maxBatchSize);

        var sendCallback = splunkBunyanStream.stream.send;
        splunkBunyanStream.stream.send = function(err, resp, body) {
            assert.strictEqual(body.text, successBody.text);
            assert.strictEqual(body.code, successBody.code);
            sendCallback(err, resp, body);
            done();
        };

        var data = "something";

        splunkBunyanStream.stream.write(formatForBunyan(data));
    });
    it("should error when writing a string with bad token", function(done) {
        var config = {
            token: "bad-token",
            maxBatchCount: 1
        };

        var splunkBunyanStream = splunkBunyan.createStream(config);

        assert.ok(splunkBunyanStream);
        assert.strictEqual("info", splunkBunyanStream.level);
        assert.strictEqual("raw", splunkBunyanStream.type);
        assert.strictEqual(config.token, splunkBunyanStream.stream.config().token);
        assert.strictEqual("splunk-bunyan-logger/0.11.0", splunkBunyanStream.stream.config().name);
        assert.strictEqual("localhost", splunkBunyanStream.stream.config().host);
        assert.strictEqual("https", splunkBunyanStream.stream.config().protocol);
        assert.strictEqual("info", splunkBunyanStream.stream.config().level);
        assert.strictEqual(8088, splunkBunyanStream.stream.config().port);

        var run = false;

        splunkBunyanStream.on("error", function(err, errContext) {
            run = true;
            assert.ok(err);
            assert.strictEqual(err.message, invalidTokenBody.text);
            assert.strictEqual(err.code, invalidTokenBody.code);
            assert.ok(errContext);

            var body = JSON.parse(errContext.message).event;
            assert.strictEqual(body.message.msg, "something");
            assert.strictEqual(body.severity, "info");
        });

        var sendCallback = splunkBunyanStream.stream.send;
        splunkBunyanStream.stream.send = function(err, resp, body) {
            assert.ok(!err);
            assert.ok(run);
            assert.strictEqual(body.text, invalidTokenBody.text);
            assert.strictEqual(body.code, invalidTokenBody.code);
            sendCallback(err, resp, body);
            unmute();
            done();
        };

        var data = "something";

        mute();
        splunkBunyanStream.stream.write(formatForBunyan(data));
    });
    it("should emit error when writing without args", function(done) {
        var config = {
            token: TOKEN,
            maxBatchCount: 1
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        splunkBunyanStream.stream.on("error", function(err) {
            assert.ok(err);
            assert.strictEqual(err.message, "Must pass a parameter to write.");
            done();
        });
        splunkBunyanStream.stream.write();
    });
    it("should not retry on Splunk error", function(done) {
        var config = {
            token: "bad-token",
            maxRetries: 5,
            maxBatchCount: 1
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        var retryCount = 0;

        // Wrap the _post so we can verify retries
        var post = splunkBunyanStream.stream.logger._post;
        splunkBunyanStream.stream.logger._post = function(requestOptions, callback) {
            retryCount++;
            post(requestOptions, callback);
        };

        splunkBunyanStream.stream.on("error", function(err) {
            assert.ok(err);
            assert.strictEqual(invalidTokenBody.code, err.code);
            assert.strictEqual(invalidTokenBody.text, err.message);
            assert.strictEqual(1, retryCount);
            unmute();
            done();
        });

        mute();
        splunkBunyanStream.stream.write("something");
    });
    it("should retry on network error, bad host", function(done) {
        this.timeout(3 * 1000);
        var config = {
            token: TOKEN,
            maxRetries: 3,
            host: "splunk.invalid",
            maxBatchCount: 1
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        var retryCount = 0;

        // Wrap the _post so we can verify retries
        var post = splunkBunyanStream.stream.logger._post;
        splunkBunyanStream.stream.logger._post = function(requestOptions, callback) {
            retryCount++;
            post(requestOptions, callback);
        };

        splunkBunyanStream.stream.on("error", function(err) {
            assert.ok(err);
            assert.strictEqual(true,["ENOTFOUND","EAI_AGAIN"].includes(err.code));
            assert.strictEqual(config.maxRetries + 1, retryCount);
            unmute();
            done();
        });

        mute();
        splunkBunyanStream.stream.write("something");
    });
    it("should retry on network error, wrong port", function(done) {
        this.timeout(3 * 1000);
        var config = {
            token: TOKEN,
            maxRetries: 3,
            port: 1075,
            maxBatchCount: 1
        };
        var splunkBunyanStream = splunkBunyan.createStream(config);

        var retryCount = 0;

        // Wrap the _post so we can verify retries
        var post = splunkBunyanStream.stream.logger._post;
        splunkBunyanStream.stream.logger._post = function(requestOptions, callback) {
            retryCount++;
            post(requestOptions, callback);
        };

        splunkBunyanStream.stream.on("error", function(err) {
            assert.ok(err);
            assert.strictEqual("ECONNREFUSED", err.code);
            assert.strictEqual(config.maxRetries + 1, retryCount);
            unmute();
            done();
        });

        mute();
        splunkBunyanStream.stream.write("something");
    });
    // TODO: add some tests for batching (interval, size, count)
});

examples/retry.js
/*
 * Copyright 2015 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/**
 * This example shows how to configure retries on errors
 * with the Splunk Bunyan logger.
 */

// Change to require("splunk-bunyan-logger");
var splunkBunyan = require("../index");
var bunyan = require("bunyan");

/**
 * Only the token property is required.
 *
 * Here we've set maxRetries to 10,
 * If there are any connection errors the request
 * to Splunk will be retried up to 10 times.
 * The default is 0.
 */
var config = {
    token: "your-token-here",
    url: "https://localhost:8088",
    maxRetries: 10
};
var splunkStream = splunkBunyan.createStream(config);

splunkStream.on("error", function(err, context) {
    // Handle errors here
    console.log("Error", err, "Context", context);
});

// Setup Bunyan, adding splunkStream to the array of streams
var Logger = bunyan.createLogger({
    name: "my logger",
    streams: [
        splunkStream
    ]
});

// Define the payload to send to HTTP Event Collector
var payload = {
    // Our important fields
    temperature: "70F",
    chickenCount: 500,

    // Special keys to specify metadata for HTTP Event Collector
    source: "chicken coop",
    sourcetype: "httpevent",
    index: "main",
    host: "farm.local"
};

// Send the payload
console.log("Sending payload", payload);
Logger.info(payload, "Chicken coup looks stable.");

examples/batching.js
/*
 * Copyright 2015 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/**
 * This example shows how to batch events with the
 * the Splunk Bunyan logger by manually calling flush.
 *
 * By default autoFlush is enabled, this means
 * an HTTP request is made for each log message.
 *
 * By disabling autoFlush, events will be queued
 * until flush() is called.
 */

// Change to require("splunk-bunyan-logger");
var splunkBunyan = require("../index");
var bunyan = require("bunyan");

/**
 * Only the token property is required.
 * 
 * Here, autoFlush is set to false
 */
var config = {
    token: "your-token-here",
    host: "localhost",
    path: "/services/collector/event/1.0",
    protocol: "https",
    port: 8088,
    level: "info",
    autoFlush: false,
    maxRetries: 0
};
var splunkStream = splunkBunyan.createStream(config);

splunkStream.on("error", function(err, context) {
    // Handle errors here
    console.log("Error", err, "Context", context);
});

// Setup Bunyan, adding splunkStream to the array of streams
var Logger = bunyan.createLogger({
    name: "my logger",
    streams: [
        splunkStream
    ]
});

// Define the payload to send to Splunk's Event Collector
var payload = {
    // Our important fields
    temperature: "70F",
    chickenCount: 500,

    // Special keys to specify metadata for Splunk's Event Collector
    source: "chicken coop",
    sourcetype: "httpevent",
    index: "main",
    host: "farm.local",
};

// Send the payload
console.log("Queuing payload", payload);
Logger.info(payload, "Chicken coup looks stable.");

var payload2 = {
    // Our important fields
    temperature: "75F",
    chickenCount: 600,

    // Special keys to specify metadata for Splunk's Event Collector
    source: "chicken coop",
    sourcetype: "httpevent",
    index: "main",
    host: "farm.local",
};

// Send the payload
console.log("Queuing second payload", payload2);
Logger.info(payload2, "New chickens have arrived");

/**
 * Since autoFlush is disabled, call flush manually.
 * This will send both payloads in a single
 * HTTP request.
 *
 * The callback for flush is optional.
 */
splunkStream.flush(function(err, resp, body) {
    // If successful, body will be { text: 'Success', code: 0 }
    console.log("Response from Splunk", body);
});


examples/basic.js
/*
 * Copyright 2015 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/**
 * This example shows basic usage of the Splunk
 * Bunyan logger.
 */

// Change to require("splunk-bunyan-logger");
var splunkBunyan = require("../index");
var bunyan = require("bunyan");

/**
 * Only the token property is required.
 */
var config = {
    token: "your-token-here",
    url: "https://localhost:8088"
};
var splunkStream = splunkBunyan.createStream(config);

splunkStream.on("error", function(err, context) {
    // Handle errors here
    console.log("Error", err, "Context", context);
});

// Setup Bunyan, adding splunkStream to the array of streams
var Logger = bunyan.createLogger({
    name: "my logger",
    streams: [
        splunkStream
    ]
});

// Define the payload to send to HTTP Event Collector
var payload = {
    // Our important fields
    temperature: "70F",
    chickenCount: 500,

    // Special keys to specify metadata for HTTP Event Collector
    source: "chicken coop",
    sourcetype: "httpevent",
    index: "main",
    host: "farm.local"
};

/**
 * Since maxBatchCount is set to 1 by default, calling send
 * will immediately send the payload.
 * 
 * The underlying HTTP POST request is made to
 *
 *     https://localhost:8088/services/collector/event/1.0
 *
 * with the following body
 *
 *     {
 *         "source": "chicken coop",
 *         "sourcetype": "httpevent",
 *         "index": "main",
 *         "host": "farm.local",
 *         "event": {
 *             "message": {
 *                 "chickenCount": 500
 *                 "msg": "Chicken coup looks stable.",
 *                 "name": "my logger",
 *                 "put": 98884,
 *                 "temperature": "70F",
 *                 "v": 0
 *             },
 *             "severity": "info"
 *         }
 *     }
 *
 */
console.log("Sending payload", payload);
Logger.info(payload, "Chicken coup looks stable.");